develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md

```diff
@@ -0,0 +1,456 @@
+
+client.Subscribe("interfaces", sampleIntervalMs, new[] { "interfaces/interface/state" }, HandleIncomingResponse);
+
+public void HandleIncomingResponse(IEnumerable<GnmiResponseValue> response)
+{
+    try
+    {
+        foreach (var value in response)
+        {
+            ProcessUpdate(value);
+        }
+    }
+    catch (Exception ex)
+    {
+        Log($"Exception occurred: {ex}");
+    }
+}
+
+
+```
+
+This will create a *SAMPLE* subscription where a notification will be sent out by the data source every 10 seconds.
+
+In case you want to do an *ON_CHANGE* subscription, you can leave out the `sampleIntervalMs` from the [Subscribe](xref:Skyline.DataMiner.Helper.OpenConfig.Api.GnmiClient.Subscribe%2A) call.
+
+```csharp
+string subscriptionName = "interfaces";
+
+client.Subscribe("interfaces", new[] { "interfaces/interface/state" }, HandleIncomingResponse);
+```
+
+Now anytime a `leaf` changes, it will send out a notification with the new value.
+
+> [!NOTE]
+> There might be some limitations on the data source in terms of granularity. For example, a switch could send out changes on the counters only once every 5 seconds while they would change multiple times in that time frame.
+
+### Accessing a specific instance
+
+If you need to access a specific instance in a `container`, you can use [ ] to specify the instance. For example: `interfaces/interface[name:Ethernet1]/state`
+
+Using this path will result in only reading or writing the *Ethernet1* instance of the *interfaces/interface/state* `container`.
+
+### Troubleshooting
+
+The errors that occur in the middleware will be logged by the `ILogger` object that gets passed along at construction of the [GnmiClient](xref:Skyline.DataMiner.Helper.OpenConfig.Api.GnmiClient).
+
+In case of a DataMiner connector, that will be the logging of the element. There is an extensive list of [error codes](xref:CommunicationGateway_ErrorCodes) available to help you troubleshoot.
+
+### DataMapper
+
+The [DataMinerConnectorDataMapper](xref:Skyline.DataMiner.Helper.OpenConfig.DataMapper.DataMinerConnectorDataMapper) is an object that sits between your device and the DataMiner connector. It will automatically parse the incoming notifications and populate DataMiner parameters or tables with the data.
+
+#### Map a YANG path to a parameter
+
+```csharp
+IDataMapper dataMapper = new DataMinerConnectorDataMapper(
+    protocol,
+    new List<IDataEntity>
+    {
+        new DataMinerConnectorParameter("system/memory/state/physical", Parameter.systemmemorystatephysical_400),
+        new DataMinerConnectorParameter("system/memory/state/reserved", Parameter.systemmemorystatereserved_404)
+    });
+```
+
+Having this configured as soon as a gNMI notification will come in with data for either the *system/memory/state/physical* or the *system/memory/state/reserved* YANG path, it will automatically do the set on the configured parameters. You need to pass it the `SLProtocol` object so it can execute those sets.
```

Comment:
This sentence is not clear. Can you rephrase this? In the very least, I'd get rid of the "Having this configured" structure and replace it with something less ambiguous.